### DISCUSSION TITLE
Mock pool for tests?

### DISCUSSION THREAD

**Discussion by karlhorky** [porsager/postgres#225] [Q&A]
Hi @porsager !

Thanks again for your continued work on this library.

Would you be open to offering a mock pool for testing database functions?

Prior art can be seen here (Slonik): https://medium.com/@gajus/mocking-postgresql-queries-in-node-js-730da98545ed

**Comment by porsager** [comment]
Hi @karlhorky 

Do you have an idea how the surface api for this should look? Maybe I don't understand the use case from slonik you linked, but it looks a bit convoluted to me.

**Comment by karlhorky** [comment]
I didn't have a specific proposal in mind - probably others who are more into this area are able to come up with something better than I can (without me doing a bunch of research as to further prior art). Looking at the Slonik example, I see what you mean about it being convoluted 🤔

Maybe others such as @throrin19, @Benricheson101, @kevinK0enig have some API design thoughts...

**Comment by throrin19** [comment]
To mock your database for tests, You have many options availables and it depends what do you want. For all, this is the available part for knex.js or other good practices (but for some, I don't know how to do that correctly in a CI/CD system)

- mock-knex : You mock all requests, test the generated sql, request bindings, ... and return what you want
- Create temporary postgres database with all your test entires in that and check results based of your requests directly. But for this example, I don't know how to do that correctly
- Use sinon.js to mock/spy/... usage of postgresjs.

**Comment by porsager** [comment]
I think I have a hard time seeing the point in mocking your database. Especially PostgreSQL which we should expect to not lie to us. Maybe if we were dealing with mongo 😋

If it's about not being able to run Postgres for the tests, I think that's the problem to solve instead of mocking it.

I'll move this issue to discussions.

**Comment by Coinhexa** [comment]
- I think the use case is to be able to test your parts of code that use postgresql via this library instead of testing the library or the underlying postgresql database itself
- In such a scenario, how do you propose testing parts of the application that make use of this library?
- I can think of 2 options: 

1. You do a unit test mock by basically stubbing functions from postgres library
2. You do integration test mocks using @testcontainers/postgres or some docker compose voodoo

- As the author of the library, perhaps your recommendation on how to go about doing both holds a lot of weight
